service_provider.go

```diff
@@ -1531,22 +1531,26 @@ func (sp *ServiceProvider) ValidateLogoutResponseRedirect(query url.Values) erro
 		return err
 	}
 
+	hasValidSignature := false
 	if query.Get("Signature") != "" && query.Get("SigAlg") != "" {
 		if err := sp.validateQuerySig(query); err != nil {
 			retErr.PrivateErr = err
 			return retErr
 		}
+		hasValidSignature = true
 	}
 
 	doc := etree.NewDocument()
-	if err := doc.ReadFromBytes(rawResponseBuf); err != nil {
+	if err := doc.ReadFromBytes(gr); err != nil {
 		retErr.PrivateErr = err
 		return retErr
 	}
 
 	if err := sp.validateSignature(doc.Root()); err != nil {
-		retErr.PrivateErr = err
-		return retErr
+		if err != errSignatureElementNotPresent || !hasValidSignature {
```

Comment:
:+1: if I understand it correctly, we're checking here if either the assertions are signed or the document itself is signed.

One of which most be true

Reply:
Yes, we only want to fire error if no signature was found at all or if it's wrong.